README.md
App-Countdown

This command line application implements "countdown", a command similar to
the UNIX sleep command only that it displays the number of seconds remaining
before the timeout.

It accepts a single command line argument of an integer with the number
of seconds left.

- [Homepage](http://www.shlomifish.org/open-source/projects/countdown/)
- [MetaCPAN Record](http://metacpan.org/release/App-Countdown)


App-Countdown/Changes
Revision history for App-Countdown

0.8.3       2020-10-07

    - Add documentation to the App::Countdown module about command-line use.
        - Thanks to lopid
    - Remove old POD sections.

0.8.2       2020-01-28
    - Rebuild for order of 'NAME' and 'VERSION' sections in the generated
    POD documentation (see:
    https://metacpan.org/changes/distribution/Pod-Weaver-PluginBundle-SHLOMIF#L3
    ).
        - VERSION used to appear before NAME.

0.8.1   2019-07-20
        - Prefer future dates and use the local time zone.
        - weaver.ini

0.8.0   2019-07-19
        - Add the --to flag.

0.6.0   2019-02-01
        - Also display the remaining delay in h:mm:ss notation.
        - Apply perltidy style.

0.4.5   2018-05-22
        - Convert to dzil .

0.4.4   2017-05-28
        - Convert the repository to git+GitHub.
            - See https://github.com/shlomif/App-Countdown .
            - This is in part to be more http://cpan-prc.org/ friendly.

0.4.3   2014-01-17
        - Hopefully fix the link to the repository resource at bitbucket.
            - It is in all lower-case.

0.4.2   2014-01-17
        - Fix a typo in the NAME and made it all on one line:
            - https://rt.cpan.org/Public/Bug/Display.html?id=92213
            - Thanks to Neil_Bowers
        - Made the minimum version of perl 5.10 in the Build.PL too.

0.4.1   2014-01-16
        - Add t/cpan-changes.t and adapted Changes to the CPAN::Changes
        convention.

0.4.0   2014-01-16
        - Implement the new syntax of "1m30s" or "1h30m50s" for delays.
        - Add examples to the bin/countdown man-page POD.
        - Add t/style-trailing-space.t .

0.3.1   2013-09-03
        - Add a minimal dependency on Getopt::Long 2.36 for
        GetOptionsFromArray().

0.3.0   2013-08-31
        - Implement support for the "s" delay.
            - it kept tripping me up, so I added it.

0.2.0   2012-11-15
        - Add --version, --help, and --man flags.
            - Thanks to Steven Haryanto from
            http://blogs.perl.org/users/shlomi_fish/2012/11/three-command-line-utilities.html#comments
        - Add t/unit-calc-delay.t
        - Implement support for m and h delays (plus fractional numbers).
            - Also thanks to Steven.

0.0.4   2012-11-13
        - URL to the repo changed to fix a typo. Updated Build.PL.

0.0.3   2012-11-13
        - Correct the phrasing in the POD of bin/countdown.
            - Thanks to Omer Zak for pointing it out.

0.0.2   2012-11-13
        - Fix a bug to accept single digit numbers for delays.

0.0.1   2012-11-13
        - First version, released on an unsuspecting world.


App-Countdown/lib/App/Countdown.pm
package App::Countdown;

use 5.010;

use strict;
use warnings FATAL => 'all';

use DateTime::Format::Natural ();
use Time::HiRes qw(sleep time);
use POSIX qw();
use IO::Handle;
use Getopt::Long qw(2.36 GetOptionsFromArray);
use Pod::Usage;
use Carp;

=head1 NAME

App::Countdown - wait some specified time while displaying the remaining time.

=head1 SYNOPSIS

    use App::Countdown ();

    App::Countdown->new({ argv => [@ARGV] })->run();

=head1 SUBROUTINES/METHODS

=head2 new

A constructor. Accepts the argv named arguments.

=head2 run

Runs the program.

=cut

sub new
{
    my $class = shift;

    my $self = bless {}, $class;

    $self->_init(@_);

    return $self;
}

sub _delay
{
    my $self = shift;

    if (@_)
    {
        $self->{_delay} = shift;
    }

    return $self->{_delay};
}

sub _end
{
    my $self = shift;

    if (@_)
    {
        $self->{_end} = shift;
    }

    return $self->{_end};
}

my $up_to_60_re = qr/[1-9]|[1-5][0-9]|0[0-9]?/;

sub _get_up_to_60_val
{
    my ($v) = @_;

    ( $v //= '' ) =~ s/\A0*//;

    return ( length($v) ? $v : 0 );
}

sub _calc_delay
{
    my ( $self, $delay_spec ) = @_;

    if ( my ( $n, $qualifier ) =
        $delay_spec =~ /\A((?:[1-9][0-9]*(?:\.\d*)?)|(?:0\.\d+))([mhs]?)\z/ )
    {
        return int(
            $n * (
                  $qualifier eq 'h' ? ( 60 * 60 )
                : $qualifier eq 'm' ? 60
                :                     1
            )
        );
    }
    elsif ( my ( $min, $sec ) =
        $delay_spec =~ /\A([1-9][0-9]*)m($up_to_60_re)s\z/ )
    {
        return $min * 60 + _get_up_to_60_val($sec);
    }
    elsif ( ( ( my $hour ), $min, $sec ) =
        $delay_spec =~
        /\A([1-9][0-9]*)h(?:($up_to_60_re)m)?(?:($up_to_60_re)s)?\z/ )
    {
        return ( ( $hour * 60 + _get_up_to_60_val($min) ) * 60 +
                _get_up_to_60_val($sec) );
    }
    else
    {
        die
"Invalid delay. Must be a positive and possibly fractional number, possibly followed by s, m, or h";
    }
}

sub _init
{
    my ( $self, $args ) = @_;

    my $argv = [ @{ $args->{argv} } ];

    my $help    = 0;
    my $man     = 0;
    my $version = 0;
    my $end_str;
    if (
        !(
            my $ret = GetOptionsFromArray(
                $argv,
                'help|h' => \$help,
                man      => \$man,
                version  => \$version,
                'to=s'   => \$end_str,
            )
        )
        )
    {
        die "GetOptions failed!";
    }

    if ($help)
    {
        pod2usage(1);
    }

    if ($man)
    {
        pod2usage( -verbose => 2 );
    }

    if ($version)
    {
        print "countdown version $App::Countdown::VERSION .\n";
        exit(0);
    }

    if ( defined $end_str )
    {
        my $parser = DateTime::Format::Natural->new(
            prefer_future => 1,
            time_zone     => 'local',
        );
        my $dt = $parser->parse_datetime($end_str);
        if ( not $parser->success )
        {
            die $parser->error;
        }
        $self->_end( $dt->epoch );
    }
    else
    {
        my $delay = shift(@$argv);

        if ( !defined $delay )
        {
            Carp::confess("You should pass a number of seconds.");
        }

        $self->_delay( $self->_calc_delay($delay) );
    }
    return;
}

sub _format
{
    my $delay = shift;
    return sprintf( "%d:%02d:%02d",
        POSIX::floor( $delay / 3600 ),
        POSIX::floor( $delay / 60 ) % 60,
        $delay % 60 );
}

sub _calc_end
{
    my ( $self, $start ) = @_;

    return defined( $self->_end ) ? $self->_end : ( $start + $self->_delay );
}

sub run
{
    my ($self) = @_;

    STDOUT->autoflush(1);

    my $start = time();
    my $end   = $self->_calc_end($start);

    my $delay = $end - $start;

    my $hms_tot = _format($delay);
    my $last_printed;
    while ( ( my $t = time() ) < $end )
    {
        my $new_to_print = POSIX::floor( $end - $t );
        if ( !defined($last_printed) or $new_to_print != $last_printed )
        {
            $last_printed = $new_to_print;
            my $hms = _format($new_to_print);
            print "Remaining $hms / $hms_tot ( $new_to_print/$delay )",
                ' ' x 10, "\r";
        }
        sleep(0.1);
    }

    return;
}

1;

=head1 USAGE

    countdown [number of seconds]
    countdown [minutes]m
    countdown [hours]h
    countdown [seconds]s
    countdown [minutes]m[seconds]s
    countdown [hours]h[minutes]m[seconds]s
    countdown --to "20:30"

=head1 OPTIONS

    --man - displays the man page.
    --help - displays the help.
    --version - displays the version.

=head1 DESCRIPTION

B<countdown> waits for a certain time to pass, in a similar fashion to the
UNIX sleep command, but unlike sleep, it displays the amount of time left to
sleep. I always found it frustrating that I've placed an alarm using
C<sleep $secs ; finish-client> and could not tell how much time left, so I
wrote B<countdown> for that.

=head1 EXAMPLES

    $ countdown 30s # 30 seconds

    $ countdown 1m  # 1 minute

    $ countdown 100 # 100 seconds

    $ countdown 2h  # 2 hours

    $ countdown 2m30s # 2 minutes and 30 seconds.

    $ countdown 1h0m30s # 1 hour, 0 minutes and 30 seconds.

=head1 AUTHOR

Shlomi Fish, L<http://www.shlomifish.org/>, C<< <shlomif at cpan.org> >> .

=head1 ACKNOWLEDGEMENTS

=over 4

=item * Neil Bowers

Reporting a typo and a problem with the description not fitting on one line.

=back

=cut

1;    # End of App::Countdown


App-Countdown/Build.PL
use 5.008;
use strict;
use warnings FATAL => 'all';

use lib './inc';

use Test::Run::Builder;

my $builder = Test::Run::Builder->new(
    module_name         => 'App::Countdown',
    license             => 'MIT',
    dist_author         => q{Shlomi Fish <shlomif@cpan.org>},
    dist_version_from   => 'lib/App/Countdown.pm',
    release_status      => 'stable',
    configure_requires => {
        'Module::Build' => 0,
    },
    build_requires => {
        'Test::More' => 0,
    },
    requires => {
        'Carp' => 0,
        'Getopt::Long' => '2.36',
        'IO::Handle' => 0,
        'perl' => '5.010',
        'POSIX' => 0,
        'strict' => 0,
        'Time::HiRes' => 0,
        'warnings' => 0,
    },
    add_to_cleanup     => [ 'App-Countdown-*' ],
    create_makefile_pl => 'traditional',
    script_files => ['bin/countdown',],
    meta_merge =>
    {
        resources =>
        {
            repository => "http://bitbucket.org/shlomif/perl-App-Countdown",
            homepage => "http://www.shlomifish.org/open-source/projects/countdown/",
        },
        keywords =>
        [
            'app',
            'application',
            'console',
            'countdown',
            'modulino',
            'sleep',
            'terminal',
            'timer',
        ],
    },
);

$builder->create_build_script();
